collector/analyzer/tools/utf8.go

```diff
@@ -0,0 +1,167 @@
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x70-0x7F
+	//   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0x80-0x8F
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0x90-0x9F
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0xA0-0xAF
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0xB0-0xBF
+	xx, xx, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, // 0xC0-0xCF
+	s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, // 0xD0-0xDF
+	s2, s3, s3, s3, s3, s3, s3, s3, s3, s3, s3, s3, s3, s4, s3, s3, // 0xE0-0xEF
+	s5, s6, s6, s6, s7, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0xF0-0xFF
+}
+
+// acceptRange gives the range of valid values for the second byte in a UTF-8
+// sequence.
+type acceptRange struct {
+	lo uint8 // lowest value for second byte.
+	hi uint8 // highest value for second byte.
+}
+
+// acceptRanges has size 16 to avoid bounds checks in the code that uses it.
+var acceptRanges = [16]acceptRange{
+	0: {locb, hicb},
+	1: {0xA0, hicb},
+	2: {locb, 0x9F},
+	3: {0x90, hicb},
+	4: {locb, 0x8F},
+}
+
+func runeIndex(p []byte) int {
+	np := len(p)
+	var n int
+	for i := 0; i < np; {
+		c := p[i]
+		if c < utf8.RuneSelf {
+			// ASCII fast path
+			if n > 0 && n < 3 {
+				return i - n
+			}
+			i++
+			n = 0
+			continue
+		}
+		x := first[c]
+		if x == xx {
+			return i - n // invalid.
+		}
+		size := int(x & 7)
+		if i+size > np {
+			if n > 0 && n < 3 {
+				return i - n
+			}
+			return i // Short or invalid.
+		}
+		accept := acceptRanges[x>>4]
+		if c := p[i+1]; c < accept.lo || accept.hi < c {
+			size = 1
+		} else if size == 2 {
+		} else if c := p[i+2]; c < locb || hicb < c {
+			size = 1
+		} else if size == 3 {
+		} else if c := p[i+3]; c < locb || hicb < c {
+			size = 1
+		}
+		n = size
+		i += size
+	}
+	return np
+}
+
+func FomratByteArrayToUtf8(p []byte) string {
+	index := runeIndex(p)
+	if index == len(p) {
+		return string(p)
+	}
+	if index < 0 {
+		return ""
+	}
+	return string(p[:index])
+}
+
+func runeIndexInString(s string) int {
+	ns := len(s)
+	var n int
+	for i := 0; i < ns; {
+		c := s[i]
+		if c < utf8.RuneSelf {
+			// ASCII fast path
+			if n > 0 && n < 3 {
+				return i - n
+			}
+			i++
+			n = 0
+			continue
+		}
+		x := first[c]
+		if x == xx {
+			return i - n // invalid.
+		}
+		size := int(x & 7)
+		if i+size > ns {
+			if n > 0 && n < 3 {
+				return i - n
+			}
+			return i // Short or invalid.
+		}
+		accept := acceptRanges[x>>4]
+		if c := s[i+1]; c < accept.lo || accept.hi < c {
+			size = 1
+		} else if size == 2 {
+		} else if c := s[i+2]; c < locb || hicb < c {
+			size = 1
+		} else if size == 3 {
+		} else if c := s[i+3]; c < locb || hicb < c {
+			size = 1
+		}
+		n = size
+		i += size
+	}
+	return ns
+}
+
+func FomratStringToUtf8(s string) string {
```

Comment:
I think `Fomrat` is a typo which is supposed to be `Format`.

Reply:
Wrong spell word, check it.

Reply:
It‘s a spelling mistake, fix it.